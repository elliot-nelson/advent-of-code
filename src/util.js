// ------------------------------------------------------------------------
// utilities for our solution modules
// ------------------------------------------------------------------------
const chalk = require('chalk');

const util = {
  log: {
    info: console.log,
    part1: value => {
      util.log.info('  Part 1: ' + chalk.green(value));
    },
    part2: value => {
      util.log.info('  Part 2: ' + chalk.green(value));
    }
  },
  sum(...args) {
    let total = 0;
    for (let arg of args) total += arg;
    return total;
  },
  fill2d(width, height, value) {
    let array = [];
    for (let y = 0; y < height; y++) {
      array[y] = [];
      for (let x = 0; x < width; x++) {
        array[y][x] = value;
      }
    }
    return array;
  }
};

module.exports = util;
