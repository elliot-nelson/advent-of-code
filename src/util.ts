// Utilities

import * as path from 'path';
import * as fs from 'fs';

export interface ISolution<T> {
    part1: T;
    part2: T;
}

export function load(...paths: string[]): string[] {
    const raw = fs.readFileSync(path.join(...paths), 'utf8');
    const lines = raw.split(/\r?\n/);

    if (lines[lines.length - 1].length === 0) {
        lines.splice(lines.length - 1, 1);
    }

    return lines;
}

export const Util = {
    group<T>(array: T[], count: number): T[][] {
        const result: T[][] = [];

        for (let i = 0; i < array.length / count; i++) {
            let k: T[] = [];
            for (let j = 0; j < count; j++) {
                k.push(array[i * count + j]);
            }
            result.push(k);
        }

        return result;
    }
};
