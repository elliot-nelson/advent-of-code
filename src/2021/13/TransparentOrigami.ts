// Year 2021 Day 13
// Transparent Origami

import { ISolution } from '../../util';

type Instruction = {
    axis: string;
    value: number;
}

function fold(dots: number[][], instr: Instruction) {
    const index = instr.axis === 'x' ? 0 : 1;

    for (const dot of dots) {
        if (dot[index] > instr.value) dot[index] = instr.value - (dot[index] - instr.value);
    }
}

function countUnique(dots: number[][]): number {
    const hash: Record<string, boolean> = dots.reduce((hash, dot) => {
        hash[String(dot)] = true;
        return hash;
    }, {});
    return Object.keys(hash).length;
}

function format(dots: number[][]): string {
    const output: string[][] = [];
    for (const dot of dots) {
        output[dot[1]] = output[dot[1]] || [];
        output[dot[1]][dot[0]] = '#';
    }
    return [...output].map(line => [...line].map(char => char || ' ').join('')).join('\n');
}

export function solve(input: string[]): ISolution<string> {
    const sections: string[] = input.join('\n').split('\n\n');
    const dots: number[][] = sections[0].split('\n').map(line => line.split(',').map(Number));
    const instructions: Instruction[] = sections[1].split('\n').map(str => {
        const elem = str.split(' ').slice(-1)[0].split('=');
        return { axis: elem[0], value: Number(elem[1]) };
    });

    // Part 1
    fold(dots, instructions[0]);
    const part1: string = String(countUnique(dots));

    // Part 2
    for (const instr of instructions.slice(1)) fold(dots, instr);
    const part2: string = format(dots);

    return { part1, part2 };
}
