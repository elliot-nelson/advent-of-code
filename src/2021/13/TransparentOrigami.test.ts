// Year 2021 Day 13
// Transparent Origami

import { load } from '../../util';
import { solve } from './TransparentOrigami';

describe('TransparentOrigami', () => {
    it('runs example 1', () => {
        const input = load(__dirname, '../../src/2021/13/', 'example1.txt');
        const solution = solve(input);

        expect(solution.part1).toEqual('17');
        expect(solution.part2).toMatchSnapshot();
    });

    it('runs my input', () => {
        const input = load(__dirname, 'input.txt');
        const solution = solve(input);

        expect(solution.part1).toEqual('710');
        expect(solution.part2).toMatchSnapshot();
    });
});
