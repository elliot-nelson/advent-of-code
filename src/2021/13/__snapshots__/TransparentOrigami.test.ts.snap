// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransparentOrigami runs example 1 1`] = `
"#####
#   #
#   #
#   #
#####"
`;

exports[`TransparentOrigami runs my input 1`] = `
"#### ###  #     ##  ###  #  # #    ###
#    #  # #    #  # #  # #  # #    #  #
###  #  # #    #    #  # #  # #    #  #
#    ###  #    # ## ###  #  # #    ###
#    #    #    #  # # #  #  # #    # #
#### #    ####  ### #  #  ##  #### #  #"
`;
